return {
	tiles = {
		Wood = {
			name = "Wood",
			description = "A block that is often accociated with building, but can also be used for low-tier weapons and armor",
			functionality = "Can be placed",
		},
		Dirt = {
			name = "Dirt",
			description = "The most dirtiest tile",
			functionality = "Can be placed",
		},
		Stone = {
			name = "Stone",
			description = "A hard block that is often used for building and crafting",
			functionality = "Can be placed",
		},
		RichMahogany = {
			name = "Rich Mahogany",
			description = "Wood commonly associated with the vast jungle trees. You can make wooden tools and armor with this",
			functionality = "Can be placed",
		},
		Ebonwood = {
			name = "Ebon Wood",
			description = "A infected piece of wood that once came from a health tree. How are you even holding this thing?",
			functionality = "Can be placed",
		},
		EbonStone = {
			name = "Ebon Stone",
			description = "A piece of stone that has been infected with the corruption. It is said that the corruption can be purified, yet no one has been able to do it yet",
			functionality = "Can be placed",
		},
		Mud = {
			name = "Mud",
			description = "Mud is oftenly found in wet areas. Both found in the Mushroom Biome and in the scarce Jungle biome",
			functionality = "Can be placed",
		}
	}
}
